adcc/AdcMatrix.py

```diff
@@ -32,6 +32,48 @@
 from .AmplitudeVector import AmplitudeVector
 
 
+class AdcExtraTerm:
+    # NOTE: currently requires the matrix
+    # to allow for block construction with the same
+    # interface as the usual adc_pp, i.e.,
+    # reference_state, ground_state, intermediates
+    def __init__(self, matrix, blocks):
+        """Initialise an AdcExtraTerm.
+        This class can be used to add customs terms
+        to an existing :py:class:`AdcMatrix`
+
+        Parameters
+        ----------
+        matrix : AdcMatrix
+            The matrix for which the extra term
+            should be created.
+        blocks : dict
+            A dictionary where the key labels the matrix block
+            and the item denotes a callable to construct
+            an :py:class:`AdcBlock`
+        """
+        self.ground_state = matrix.ground_state
+        self.reference_state = matrix.reference_state
+        self.intermediates = matrix.intermediates
+        self.blocks = {}
+        if not isinstance(blocks, dict):
+            raise TypeError("blocks needs to be a dict.")
+        for space in blocks:
+            block_fun = blocks[space]
+            if not callable(block_fun):
+                raise TypeError("Items in additional_blocks must be callable.")
+            block = block_fun(
+                self.reference_state, self.ground_state, self.intermediates
+            )
+            self.blocks[space] = block
+
+    def __iadd__(self, other):
+        # disable in-place addition getting accidentally
+        # short-circuited to AdcMatrix.__radd__
+        raise NotImplementedError("In-place addition not implemented "
+                                  " for AdcExtraTerm")
```

Comment:
This is why Julia is so much nicer ...

Reply:
Yes... I agree haha

Reply:
oh also this is again a case where you raise NotImplemented and not an error message

Reply:
Done.